shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -81,6 +126,19 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
     private def containsMth(key: (Str, Str) \/ (Opt[Str], Str)): Bool = mthEnv.contains(key) || parent.exists(_.containsMth(key))
     def containsMth(parent: Opt[Str], nme: Str): Bool = containsMth(R(parent, nme))
     def nest: Ctx = copy(Some(this), MutMap.empty, MutMap.empty)
+    /**
+      * Enter a new quoted environment with empty `quoteSkolemEnv` and `freeVarsInCurrentQuote`.
+      * A whole quasiquote (i.e., code"...") contains no binding or free variables at the beginning.
+      * For a quoted binding (e.g., code"x => ..."):
+      *  1. The context of the lambda body is still in the quotation so `inQuote = true`
+      *  2. `quoteSkolemEnv` only contains skolems created by the current binding and `freeVarsInCurrentQuote` only contains free variables appearing in the body.
+      * So we also apply empty `quoteSkolemEnv` and `freeVarsInCurrentQuote` at the beginning.
+      * e.g. `code"x => y => x + y"`. For `y => x + y`, freeVarsInCurrentQuote = {'gx, 'gy}, quoteSkolemEnv = {'gy}.
+      * To get the contextual type, we solve constraint the 'gx \/ 'gy <= 'a \/ 'gy, where 'a is a fresh type variable for `y => x + y`'s contextual type.
+      * So for `code"x => ..."`, freeVarsInCurrentQuote = {'a}, quoteSkolemEnv = {'gx}, where 'gx <= 'a.
+      */
+    def enterQuotedScope: Ctx = copy(Some(this), MutMap.empty, MutMap.empty, inQuote = true, quoteSkolemEnv = MutMap.empty, freeVarsInCurrentQuote = MutSet.empty)
```

Comment:
A few remarks:
- It seems `nest` should never be called while inside a quote; instead, `enterQuotedScope` should be called instead. Is that right? If so, add an assertion that it is indeed the case in `nest`.
- It seems every use of `enterQuotedScope` should be paired with some code to handle free variable uses at the point the scope is discarded. Is that right? If so, document it – it's a very subtle invariant that's easy to get wrong, hence it should be explained and not left implicit!

Reply:
I've added the assertion and documented the precausions for `enterQuotedScope`.